xarray/core/resample.py

```diff
@@ -43,167 +67,157 @@ def _flox_reduce(self, dim, **kwargs):
         labels = np.repeat(self._unique_coord.data, repeats)
         group = DataArray(labels, dims=(self._group_dim,), name=self._unique_coord.name)
 
-        result = super()._flox_reduce(dim=dim, group=group, **kwargs)
+        result = super()._flox_reduce(
+            dim=dim, group=group, keep_attrs=keep_attrs, **kwargs
+        )
         result = self._maybe_restore_empty_groups(result)
         result = result.rename({RESAMPLE_DIM: self._group_dim})
         return result
 
-    def _upsample(self, method, *args, **kwargs):
-        """Dispatch function to call appropriate up-sampling methods on
-        data.
-
-        This method should not be called directly; instead, use one of the
-        wrapper functions supplied by `Resample`.
-
-        Parameters
-        ----------
-        method : {"asfreq", "pad", "ffill", "backfill", "bfill", "nearest", \
-                 "interpolate"}
-            Method to use for up-sampling
-
-        See Also
-        --------
-        Resample.asfreq
-        Resample.pad
-        Resample.backfill
-        Resample.interpolate
-
-        """
-
-        upsampled_index = self._full_index
-
-        # Drop non-dimension coordinates along the resampled dimension
-        for k, v in self._obj.coords.items():
-            if k == self._dim:
-                continue
-            if self._dim in v.dims:
-                self._obj = self._obj.drop_vars(k)
-
-        if method == "asfreq":
-            return self.mean(self._dim)
-
-        elif method in ["pad", "ffill", "backfill", "bfill", "nearest"]:
-            kwargs = kwargs.copy()
-            kwargs.update(**{self._dim: upsampled_index})
-            return self._obj.reindex(method=method, *args, **kwargs)
-
-        elif method == "interpolate":
-            return self._interpolate(*args, **kwargs)
-
-        else:
-            raise ValueError(
-                'Specified method was "{}" but must be one of'
-                '"asfreq", "ffill", "bfill", or "interpolate"'.format(method)
-            )
+    def _drop_coords(self) -> T_Xarray:
+        """Drop non-dimension coordinates along the resampled dimension."""
+        obj = self._obj
+        for k, v in obj.coords.items():
+            if k != self._dim and self._dim in v.dims:
+                obj = obj.drop_vars(k)
+        return obj
 
-    def asfreq(self):
+    def asfreq(self) -> T_Xarray:
```

Comment:
If you move `asfreq` down to `DataArrayResample` and `DatasetResample` then the `# type: ignore[attr-defined]` shouldn't be necessary I think,